### DISCUSSION TITLE
Really struggling to get car charging to work (GivEnergy EVC)

### DISCUSSION THREAD

**Discussion by DJBenson** [springfall2008/batpred#1248] [General]
I've just added a GivEnergy EVC to my setup and I'm really struggling to get the car charging to work reliably (if at all).

In the UI I have the following set;
- Car Charging Hold (On)
- Car charging manual SOC (On)
- Car Charging Plan Smart (On)
- Car Charging Plan max price (20p)
- Car charging planned ready time (08:00:00)

In apps.yaml I have the following;
- num_cars (1)
- car_charging_energy (sensor.givenergy_ev_energy) | Current value: 79.81 (this value is retrieved from the GivEnergy API using a restful sensor)
- car_charging_planned (binary_sensor.givenergy_ev_charger_connected) | Current value: true (this value is retrieved from the GivEnergy API using a restful sensor)
- car_charging_planned_response contains all possible options for a boolean response
- car_charging_now (sensor.givenergy_ev_status) | Current value: Preparing (this value is retrieved from the GivEnergy API using a restful sensor)
- car_charging_now_response - contains "charging" which is the coorect status from the EVC
- car_charging_battery_size - 78
- car_charging_limit (number.eevee_ac_charging_limit) | Current value: 100 (this value is set by an input number provided by the Kia integration)
- car_charging_soc (sensor.eevee_ev_battery_level) | Current value: 64 (this value is a sensor provided by the Kia integration)

The plan just seems to never include the car.

For the avoidance of doubt - all input values are from sensors which are available (i.e. they've not gone "unavailable" and thus returning invalid values).

Below is an example of the CAR0 lines from the log;

`2024-06-23 20:30:32.756689:    CAR0: [  128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21,   128.21]`

...and the only other mention of car in the logs...

`2024-06-23 20:30:01.539183: Cars 1 charging from battery False planned [False], charging_now [False] smart [True], max_price [20.0], plan_time ['08:00:00'], battery size [78.0], limit [78.0], rate [7.4]
`
As you can (hopefully) see, there is no car charging planned. As the SOC is (64) is less than the maximum (100) and the unit rate (20) is less than many of the rates before 8am, I would expect to see charging planned.

<img width="347" alt="image" src="https://github.com/springfall2008/batpred/assets/1013909/54e5df74-afa1-446d-91da-d2927ada2760">

**Comment by gcoan** [comment]
Can you share your apps.yaml and logfile please.  I agree, it looks weird, should be planning charging.

You mentioned the sensor settings you have set in apps.yaml, there are some HA configuration items that have to be set, things like the time the car charging is needed by

Can you check these are all set correctly https://springfall2008.github.io/batpred/car-charge-planning/

**Reply by DJBenson** [reply]
The settings in Home Assistant are already listed above. 

I'll add the requested files later. 

Thanks for the response.

**Reply by gcoan** [reply]
> The settings in Home Assistant are already listed above.

Oh yes you did, sorry.  They all look fine.  Car charging should be planned

**Reply by DJBenson** [reply]
"Should" being the operative word. I just can't see why it's not. No errors, just not wanting to schedule any slots.

**Comment by DJBenson** [comment]
@springfall2008 do you have any (further) thoughts on this?

I've stripped the car config back to the barebones, removing any optional config and it still doesn't plan so I'm lost.

**Reply by gcoan** [reply]
I don't have an EV so am shooting in the dark a bit and guessing.

In your settings:

> Car Charging Hold (On)
> car_charging_energy (sensor.givenergy_ev_energy) | Current value: 79.81 (this value is retrieved from the GivEnergy API using a restful sensor)
car_charging_planned 

I notice this is slightly above 

> car_charging_battery_size - 78

Wondering if this is causing confusion as 79.81 > 78?  It shouldn't because the first two are about filtering out historical car charging data from your house load not the battery SoC, but maybe turn car charging hold off?

The other thing I was wondering about was 

> Below is an example of the CAR0 lines from the log;
> 
> 2024-06-23 20:30:32.756689: CAR0: [ 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21, 128.21,

This is the projected car SoC in kWh and is flat, but moreover it is higher than the battery SoC % figure should have arrived at.

I'm wondering if somewhere there is a misconfiguration between battery capacity in kWh and current SoC in % so it believes your car battery is already too full?

**Reply by DJBenson** [reply]
sensor.givenergy_ev_energy is a cumulative sensor which tells me (and Predbat) how much energy has been put into the car (and as such can be removed from any calculations as it's not "house load"). It's not related to car_charging_battery_size to the best of my knowledge. The energy sensor will now always be above the battery capacity.

The only possibility I can think of for a mismatch is on percentages. My Kia integration represents SoC as a percentage as an integer - is this what Predbat expects or does it expect 0.64 instead of 64?

I've honestly taken it apart and put it back together so many times I just can't work out what's going wrong. It did work very intermitently but then stopped, which may have coincided with the energy sensor exceeding the battery capacity value but I cannot for the life of me think why that's remotely relevant.

Will try disabling car charging hold...

**Reply by DJBenson** [reply]
This is really confusing;
![image](https://github.com/springfall2008/batpred/assets/1013909/73be5f92-9c46-4160-8024-29b9ad2adf3b)

Is Predbat calculating "the kWh currently left in the car" based on the SoC and the battery size? The two values are not comparable (percentage versus kWh) unless there is a conversion behind the scenes.

Assuming it does, at the time I created this question, the following was true;
- SoC (%) - 64
- Battery capacity (kWh) - 78
- Estimated battery SoC (kWh) - 49.92 (64 / 100 * 78)

So in two cases where a comparison does make sense;
- SoC (64 vs 100) and
- SoC as kWh (49.92 vs 78)

Both satisfy the condition "the car should be charged". I just can't see why there is seemingly a comparison between percentage and SoC.

**Reply by DJBenson** [reply]
I decided to try something - I turned off "Car charging manual SOC" and now the CAR0 lines are reading 97.

`2024-06-26 14:50:11.744231:    CAR0: [    97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0,     97.0]`

97 is the current SoC as reported by the Kia integration...

![image](https://github.com/springfall2008/batpred/assets/1013909/4df8fb41-22d6-4d05-be9d-2fe273bd9f43)

**Reply by DJBenson** [reply]
OK I think I understand the calculations a little better. Based on this line;

`2024-06-26 15:03:02.045538: Cars 1 charging from battery False planned [False], charging_now [False] smart [True], max_price [25.0], plan_time ['08:00:00'], battery size [77.4], limit [69.66], rate [7.4]`

The limit is a calculation of the target % entity as a percentage of the battery capacity - I changed the value to 90 and saw the "limit" change from 77.4 (when it was at 100%) to 69.66 which is 90% of 77.4. So internally Predbat is using kWh values for "what's in there currently" and "what needs to be there at the end of the plan".

I suspect the spurious value for SoC was causing this but I've no idea where it was pulled from. @springfall2008 when "Car charging manual SOC" is enabled - what does it do and where does the value (128.21) I saw come from?

**Reply by gcoan** [reply]
> I suspect the spurious value for SoC was causing this but I've no idea where it was pulled from. @springfall2008 when "Car charging manual SOC" is enabled - what does it do and where does the value (128.21) I saw come from?

This is covered in the documentation under customisation https://springfall2008.github.io/batpred/customisation/#car-charging if you set this switch to ON then it ignores the car SoC reading and uses  input_number.predbat_car_charging_manual_soc_kwh to hold the SoC value - its used when your car can't report its SoC via the EV charger

This may be the problem?

And with regard to your earlier question about SoC and kWh, Predbat expects SoC as an integer so "64", and yes it converts between SoC and car capacity to work out the current car kWh value.

Was wondering if rather than you wiring all the EV charger sensors directly into predbat you created some helper entities of the right types and then you can manually change to work out your problem, eg soc, car capacity etc

**Reply by DJBenson** [reply]
It partially answers the question - however that input number was set to 100 so I'm still unclear where the 128.21 came from. What I have learned is that the switch should be off in my case. I think I was told to turn it on but let's put that one down to RTFM.

I'm hoping the issue is resolved now but can't test yet as my wife has taken the car.

**Reply by gcoan** [reply]
I will look to tidy up the documentation, make the manual soc more clear and move it from customisation into the car charging doc so it's less spread out

**Reply by DJBenson** [reply]
Thanks for your help @gcoan - will report back later as to whether it's resolved or not.

**Reply by DJBenson** [reply]
Here we go 😎

![image](https://github.com/springfall2008/batpred/assets/1013909/98555db6-bc9a-48dc-90f4-1c5e94fd2ee1)

**Reply by DJBenson** [reply]
![image](https://github.com/springfall2008/batpred/assets/1013909/9c360dbe-480e-4ac5-ab14-89192f151269)

**Reply by gcoan** [reply]
Looks like it charged for 15 minutes then 10 minutes.  Is that what you were expecting?

**Reply by DJBenson** [reply]
I had “plug and go” turned on so my charger kept forcing itself to charge  whilst Predbat was trying to stop it 🤣

Turned it off and it’s now scheduled to charge in the early hours which correlates with the cheapest rates.

**Reply by gcoan** [reply]
Party, party !

**Reply by DJBenson** [reply]
Just a follow up - I found another issue with my setup (which may affect other GivEVC users using the RESTful integration) - the sensor for accumulated energy has a low refresh time as it's not considered important enough to refresh more often. On restarting of HA (or refreshing of the RESTful integration) the sensor goes "unknown" which Predbat does not like.

There's two fixes for this - increase the refresh time so the value is polled more often or create a template sensor which is only updated on a valid value - I did the latter.

```
- trigger:
    - platform: state
      entity_id: sensor.givenergy_ev_energy
      not_to: 
        - unknown
        - unavailable
   
  sensor:
  - name: "GivEnergy EV Energy Template"
    unique_id: 5f785f50-1fbf-4d5d-87b3-d3970a8a53d1
    state: '{{ trigger.to_state.state }}'
    device_class: "energy"
    state_class: "total_increasing"
    unit_of_measurement: "kWh"
```

The template sensor (which I've now plugged into Predbat) is always available and only gets updated when the parent sensor provides a valid value. When the parent goes unknown, the template simply retains its previous value.

**Comment by gcoan** [comment]
I've been working tonight on rewriting and simplifying the car charging documentation for Predbat, moving information out of the Customisation document into Car Charging

You can see the draft of that revised document here https://github.com/gcoan/batpred/blob/main/docs/car-charging.md

**Reply by isteel** [reply]
Hi,
I'm probably a bit too late to this but I've failed miserably to get car charging to work. I've followed the documentation (I think) but no matter what I do I either get a very small charge planned or it maxs out to filling the car battery totally.
Would it be possible to provide a simple list of what config variables are mandatory, optional, whether they hold a % or kWh value, and a brief description of what they are used for?
I've resorted to digging through the code (I'm a programmer but not a python one) to try to work out what they are used for but am still making no headway.
I'm now in the position of abandoning using predbat for car charging and simply setting a schedule manually on the charger - this will of course screw up predbat's plan as there will suddenly be a massive demand for electricity.
Thanks

**Reply by gcoan** [reply]
> Would it be possible to provide a simple list of what config variables are mandatory, optional, whether they hold a % or kWh value, and a brief description of what they are used for?

I think all the information you are looking for is already in the Predbat documentation https://springfall2008.github.io/batpred/car-charging/

Plenty of other people have got car charging working correctly, so it's probably something simple wrong with your setup.

Can you describe in more detail your setup, assume it is Predbat-led charging not IOG?   Share the appropriate bits of your apps.yaml and the predbat configuration for car charging.

Are you able to manually start and stop the car charging from home assistant, e.g. by running the start and stop commands in your car charging automation?  This is the first key step, if you can't do this then the problem lies outside Predbat.

**Reply by isteel** [reply]
I appreciate others have it working and it's purely down to my configuration. The issue is I can't determine what the variables are used for.
For example the documentation example, with regard to the MG, says to create 'car_manual_soc' but I can't see any reference to where it should be used. Or does it have to be named exactly as stated and in which case predbat will use it automatically? I searched through the predbat code for 'car_manual_soc' and it only appears in the documentation.

Yes I've got the integration from HA to charger working. Previously I was using a handrolled solution which chosen the cheapest Agile slots and started and stopped the Zappi charger.
The issue is that I can't understand how to configure predbat with regard to soc settings. This is resulting in Predbat either creating insufficient charging slots or way too many.

I'm going to delete all the car charging settings and start from scratch, again, to see if that fixes it.

**Reply by isteel** [reply]
An update:

I followed the instructions (after deleting and resetting the car charging in apps.yaml) in your documentation here 
https://github.com/gcoan/batpred/blob/main/docs/car-charging.md 
and it appears to be working much more accurately. 😊

Thanks for providing this - much relief!

**Reply by gcoan** [reply]
No problem. I did do a pretty big rewrite of the car charging documentation so hopefully it is much clearer and the sensors are explained as to what they do.

That fork of the documentation has been merged into the main Predbat codebase as well.

I did a quick read through and the stuff about car_manual_soc looks OK.  I did see that I've been a bit inconsistent in some of the switch settings, sometimes saying to turn On or Off, sometimes saying True or False, but whilst the language is inconsistent I think the intent is still clear 😃

**Reply by gcoan** [reply]
> I did see that I've been a bit inconsistent in some of the switch settings, sometimes saying to turn On or Off, sometimes saying True or False, but whilst the language is inconsistent I think the intent is still clear 😃

Just to close this off, have updated the car charging and customisation documentation to use On/Off terminology consistently throughout as that is what the HA switches are set to

**Comment by springfall2008** [comment]
>Car charging manual SOC (On)

This means you have to reset the car SOC yourself to the current car battery level before each charge, using the number sensor Predbat creates. Did you forget to do that? If you have a car_charging_soc sensor already then turn this option off.

**Reply by isteel** [reply]
I'd got a mixture of soc kWh and soc % entities from trying to get it work that it was easier to just delete everything and start again. All sorted now and worked well over night.
Thanks.

**Comment by springfall2008** [comment]
Great :)
